indico/modules/events/registration/models/registrations.py

```diff
@@ -323,33 +323,36 @@ def merge_users(cls, target, source):
                 r.user = target
 
     @hybrid_method
-    def is_publishable(self, is_participant):
+    def is_publishable(self, user_regform, hide_participants_from_other_forms=False):
         if self.visibility == RegistrationVisibility.nobody or not self.is_state_publishable:
             return False
         if (self.registration_form.publish_registrations_duration is not None
                 and self.event.end_dt + self.registration_form.publish_registrations_duration <= now_utc()):
             return False
         if self.visibility == RegistrationVisibility.participants:
-            return is_participant
+            return self.registration_form == user_regform \
+                if hide_participants_from_other_forms else user_regform is not None
```

Comment:
I haven't properly tested this yet but the logic looks wrong to me - does this work for a scenario in which a user is registered in more than one regform?

it seems to me the correct logic for this should be to pass the user to this method and then check if they are a participant in the regform

Reply:
'user_regform' has value only if user is registered in a regform of the same event.
The simplification is we consider only one regform in rare case when user had registered in several as mentioned here:
https://github.com/indico/indico/pull/6176/files#diff-5c4b5cd0ee2d617abc1f4777e9ffe30462444417a1e9ee2d3bc45f271c05d8a8

So the logic is correct considering the hypothesis.

Reply:
that comment refers only to the fact that we don't de-duplicate the list of participants when showing it in an event header, which normally only affects meetings and lectures. there, yes, it is very rare that a participant is registered in multiple regforms. but for conferences it's entirely different, and this bug would also affect conferences

moreover, the level of severity of seeing a participant twice vs. not seeing participants when you're supposed to is quite different. thus I don't think this simplification should be applied here

Reply:
Code updated to deal with case user registered in multiple forms.
I will add some more tests once we agreed on "is_publishable" logic (test for the new behaviour requires a 2nd dummy reg).

Reply:
I've tested a bit and it seems to behave well! I will review the code more closely this afternoon and I'll get back to you

For the tests, I think the simplest way to create registrations is using `indico.modules.events.registration.util.create_registration` directly, like in `registration/util_test.py`. There is also a `create_registration` fixture, but that one requires creating a user each time...